File: answers/leetcode/Minimum Depth of Binary Tree/Minimum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        queue = [root]
        min_depth = -1
        while len(queue):
            min_depth += 1
            for i in range(len(queue)):
                current = queue.pop(0)
                if current:
                    if None == current.left and None == current.right:
                        return min_depth + 1
                    queue.append(current.left)
                    queue.append(current.right)
        return min_depth

File: answers/leetcode/Minimum Depth of Binary Tree/Minimum Depth of Binary Tree.cpp
// accepted 2014-03-21
//@result 41 / 41 test cases passed. Status: Accepted Runtime: 12 ms Submitted: 0 minutes ago You are here! Your runtime beats 5.11% of cpp submissions.

class Solution {
public:
	int minDepth(TreeNode *root) {
		if (NULL == root)
		{
			return 0;
		}
		if (NULL == root->left)
		{
			return 1 + minDepth(root->right);
		}
		if (NULL == root->right)
		{
			return 1 + minDepth(root->left);
		}
		int left = minDepth(root->left);
		int right = minDepth(root->right);
		return 1 + (left < right ? left : right);
	}
};
